curvefs/src/metaserver/copyset/concurrent_apply_queue.h

```diff
@@ -0,0 +1,141 @@
+#include <utility>
+
+#include "include/curve_compiler_specific.h"
+#include "src/common/concurrent/count_down_event.h"
+#include "src/common/concurrent/task_queue.h"
+#include "curvefs/src/metaserver/copyset/operator_type.h"
+
+using curve::common::CountDownEvent;
+
+namespace curvefs {
+namespace metaserver {
+namespace copyset {
+
+using ::curve::common::GenericTaskQueue;
+
+struct ConcurrentApplyOption {
+    int wconcurrentsize;
+    int wqueuedepth;
+    int rconcurrentsize;
+    int rqueuedepth;
+};
+
+enum class ThreadPoolType {READ, WRITE};
+
+class CURVE_CACHELINE_ALIGNMENT ConcurrentApplyModule {
+ public:
+    ConcurrentApplyModule(): start_(false),
+                             rconcurrentsize_(0),
+                             rqueuedepth_(0),
+                             wconcurrentsize_(0),
+                             wqueuedepth_(0),
+                             cond_(0) {}
+
+    /**
+     * Init: initialize ConcurrentApplyModule
+     * @param[in] wconcurrentsize: num of write threads
+     * @param[in] wqueuedepth: depth of write queue in ervery thread
+     * @param[in] rconcurrentsizee: num of read threads
+     * @param[in] wqueuedephth: depth of read queue in every thread
+     */
+    bool Init(const ConcurrentApplyOption &opt);
+
+    /**
+     * Push: apply task will be push to ConcurrentApplyModule
+     * @param[in] key: used to hash task to specified queue
+     * @param[in] optype: operation type defined in proto
+     * @param[in] f: task
+     * @param[in] args: param to excute task
+     */
+    template <class F, class... Args>
+    bool Push(uint64_t key, OperatorType optype, F&& f, Args&&... args) {
+        switch (Schedule(optype)) {
+            case ThreadPoolType::READ:
+                rapplyMap_[Hash(key, rconcurrentsize_)]->tq.Push(
+                        std::forward<F>(f), std::forward<Args>(args)...);
+                break;
+            case ThreadPoolType::WRITE:
+                wapplyMap_[Hash(key, wconcurrentsize_)]->tq.Push(
+                        std::forward<F>(f), std::forward<Args>(args)...);
+                break;
+        }
+
+        return true;
+    }
+
+    /**
+     * Flush: finish all task in write threads
+     */
+    void Flush();
+    void FlushAll();
+
+    void Stop();
+
+ private:
+    bool checkOptAndInit(const ConcurrentApplyOption &option);
+
+    void Run(ThreadPoolType type, int index);
+
+    static ThreadPoolType Schedule(OperatorType optype);
+
+    void InitThreadPool(ThreadPoolType type, int concorrent, int depth);
+
+    static int Hash(uint64_t key, int concurrent) {
+        return key % concurrent;
+    }
+
+ private:
+    struct TaskThread {
+        std::thread th;
+        GenericTaskQueue<bthread::Mutex, bthread::ConditionVariable> tq;
+        explicit TaskThread(size_t capacity) : tq(capacity) {}
+    };
+
+    std::atomic<bool> start_;
+    int rconcurrentsize_;
+    int rqueuedepth_;
+    int wconcurrentsize_;
+    int wqueuedepth_;
```

Comment:
just use apply option instead of extract them.

Reply:
if do that, the code in `concurrent_apply_queue` will have many `options.xxx`, seem more ugly than this way.